### PR TITLE
trigger review for users who get auto-rate-limited

### PR DESCRIPTION
**Pull request by b0b3rt**
We want to see when users get auto-rate-limited.  Put users who get auto-rate-limited, or who get a stricter rate limit applied to them, into the review queue.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205228334715908) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/callbacks/votingCallbacks.ts [hunk 1]
```diff
@@ -14,12 +14,15 @@ import { forumTypeSetting } from '../../lib/instanceSettings';
...
+voteCallbacks.castVoteAsync.add(async function updateKarma({newDocument, vote}: VoteDocTuple, collection: CollectionBase<DbVoteableType>, user: DbUser, context) {
   // Only update user karma if the operation isn't done by one of the item's current authors.
   // We don't want to let any of the authors give themselves or another author karma for this item.
+  // We need to await it so that the subsequent check for whether any stricter rate limits apply can do a proper comparison between old and new karma
   if (!vote.authorIds.includes(vote.userId) && collectionsThatAffectKarma.includes(vote.collectionName)) {
-    void Users.rawUpdateMany({_id: {$in: vote.authorIds}}, {$inc: {karma: vote.power}});
+    await Users.rawUpdateMany({_id: {$in: vote.authorIds}}, {$inc: {karma: vote.power}});
   }
+
+  void checkForStricterRateLimits(vote.authorIds, context);
```
Comment (b0b3rt) [inline packages/lesswrong/server/callbacks/votingCallbacks.ts:25]: This is the entry-point for the new logic.

File: packages/lesswrong/server/rateLimitUtils.ts [hunk 2]
```diff
@@ -127,17 +133,21 @@ async function getCommentsOnOthersPosts(comments: Array<DbComment>, userId: stri
...
   modPostSpecificRateLimitHours: number,
   userCommentRateLimit: UserRateLimit<'allComments'> | null,
-  postId: string | null
+  postId: string | null,
+  recentKarmaInfo: RecentKarmaInfo,
   context: ResolverContext
 }): Promise<Array<RateLimitInfo>> {
-  const userIsAuthor = await getUserIsAuthor(user._id, postId, context)
-  const commentsOnOthersPostsInTimeframe =  await getCommentsOnOthersPosts(commentsInTimeframe, user._id)
+  const [userIsAuthor, commentsOnOthersPostsInTimeframe] = await Promise.all([
```
Comment (b0b3rt) [inline packages/lesswrong/server/rateLimitUtils.ts:146]: Just some performance improvement by shooting off both things at once

File: packages/lesswrong/server/rateLimitUtils.ts [hunk 3]
```diff
@@ -162,9 +171,11 @@ export async function rateLimitDateWhenUserNextAbleToPost(user: DbUser): Promise
   if (shouldIgnorePostRateLimit(user)) return null;
   
   // does the user have a moderator-assigned rate limit?
-  const [modRateLimitHours, userPostRateLimit] = await Promise.all([
+  // also get the recent karma info, we'll need it later
+  const [modRateLimitHours, userPostRateLimit, recentKarmaInfo] = await Promise.all([
     getModRateLimitHours(user._id),
-    getUserRateLimit(user._id, 'allPosts')
+    getUserRateLimit(user._id, 'allPosts'),
+    getRecentKarmaInfo(user._id)
```
Comment (b0b3rt) [inline packages/lesswrong/server/rateLimitUtils.ts:178]: Just some performance improvement by getting the recentKarmaInfo at the same time as everything else

File: packages/lesswrong/server/rateLimitUtils.ts [hunk 4]
```diff
@@ -185,10 +196,12 @@ export async function rateLimitDateWhenUserNextAbleToComment(user: DbUser, postI
...
 
   // does the user have a moderator-assigned rate limit?
-  const [modRateLimitHours, modPostSpecificRateLimitHours, userCommentRateLimit] = await Promise.all([
+  // also get the recent karma info, we'll need it later
+  const [modRateLimitHours, modPostSpecificRateLimitHours, userCommentRateLimit, recentKarmaInfo] = await Promise.all([
     getModRateLimitHours(user._id),
     getModPostSpecificRateLimitHours(user._id),
-    getUserRateLimit(user._id, 'allComments')
+    getUserRateLimit(user._id, 'allComments'),
+    getRecentKarmaInfo(user._id)
```
Comment (b0b3rt) [inline packages/lesswrong/server/rateLimitUtils.ts:204]: Just some performance improvement by getting the recentKarmaInfo at the same time as everything else

File: packages/lesswrong/server/repos/VotesRepo.ts [hunk 5]
```diff
@@ -186,17 +186,30 @@ export default class VotesRepo extends AbstractRepo<DbVote> {
...
   // (note: needs to get the user's own self-upvotes so that
   // it doesn't skip posts with no other votes)
   async getVotesOnRecentContent(userId: string): Promise<RecentVoteInfo[]> {
-    const voteFields = `"Votes"._id, "Votes"."userId", "Votes"."power", "Votes"."documentId", "Votes"."collectionName"`
     const votes = await this.getRawDb().any(`
       (
-        SELECT ${voteFields}, "Posts"."postedAt", "Posts"."baseScore" AS "totalDocumentKarma"
-        FROM "Votes"
-        JOIN "Posts" on "Posts"._id = "Votes"."documentId"
+        ${this.selectVotesOnPostsJoin}
```
Comment (b0b3rt) [inline packages/lesswrong/server/repos/VotesRepo.ts:212]: Just pulling out some repeated bits.

File: packages/lesswrong/lib/rateLimits/utils.ts [hunk 6]
```diff
@@ -158,7 +160,7 @@ function getRateLimitName (rateLimit: AutoRateLimit) {
   return rateLimitName += ` (${thresholdInfo.join(", ")})`
 }
 
-function getActiveRateLimits (user: SunshineUsersList, autoRateLimits: AutoRateLimit[]) {
+function getActiveRateLimits<T extends AutoRateLimit>(user: UserKarmaInfo & { recentKarmaInfo: RecentKarmaInfo }, autoRateLimits: T[]) {
```
Comment (b0b3rt) [inline packages/lesswrong/lib/rateLimits/utils.ts:163]: Tightening up the input and preserves the type of rate limit for the output

**Comment by Raemon** [comment]
Things Robert tested:
– Checked that it correctly put a user into review queue when going from default rate limits to being a negative karma user
– Checked that it correctly put a user into review queue when getting significantly negative rate limited
– Checked that it didn't put a user into review queue when going down in rate limit strictness.

**Review by Raemon — APPROVED** [review]

File: packages/lesswrong/server/rateLimitUtils.ts [hunk 7]
```diff
@@ -216,3 +232,75 @@ export async function getRecentKarmaInfo (userId: string): Promise<RecentKarmaIn
...
+
+export async function checkForStricterRateLimits(userIds: string[], context: ResolverContext) {
+  // We can't use a loader here because we need the user's karma which was just updated by this vote
+  const votedOnUsers = await Users.find({ _id: { $in: userIds } }).fetch();
+
+  const votesRepo = new VotesRepo();
+
+  void Promise.all(votedOnUsers.map(async (user) => {
+    const allVotes = await votesRepo.getVotesOnRecentContent(user._id);
+    console.log({ allVotes });
```
Comment (jpaddison3) [inline packages/lesswrong/server/rateLimitUtils.ts:298]: lint error is real

Comment (b0b3rt) [reply packages/lesswrong/server/rateLimitUtils.ts]: Yeah, sorry, was just debugging the integration tests (which, turns out, actually caught real issues!)

**Comment by b0b3rt** [comment]
@jpaddison3 I've also gone ahead and forum-gated this, since the EA forum only has 2 automatic rate limits and I'd be a bit surprised if you wanted users to show up in the review queue if they make 4 comments in 30 minutes with <30 karma (which is one of them).

**Pull request merged**